src/Uno.UI/UI/Xaml/Controls/TextBlock/TextBlock.cs

```diff
@@ -393,6 +393,71 @@ Brush Foreground
 				)
 			);
 
+
+		private void SetDefaultForeground()
+		{
+			// override the default value from dependency property based on actual theme
+			if (ActualTheme == ElementTheme.Default)
+			{
+				throw new InvalidOperationException("Actual theme may not be default");
+			}
+
+			// find nearest parent with set Foreground or RequestedTheme
+			object candidate = this;
+			while (candidate != null)
+			{
+				if (candidate is FrameworkElement candidateFe)
+				{
+					if (candidateFe.RequestedTheme != ElementTheme.Default ||
+						candidateFe.GetCurrentHighestValuePrecedence(Control.ForegroundProperty) < DependencyPropertyValuePrecedences.Inheritance)
+					{
+						break;
+					}
+				}
+				candidate = candidate?.GetParent();
+			}			
+
+			if (candidate != null)
+			{
+				var foregroundSource = (FrameworkElement)candidate;
+				if (foregroundSource.RequestedTheme != ElementTheme.Default)
+				{
+					this.SetValue(
+						ForegroundProperty,
+						foregroundSource.RequestedTheme == ElementTheme.Light ?
+							SolidColorBrushHelper.Black : SolidColorBrushHelper.White,
+						DependencyPropertyValuePrecedences.ImplicitStyle);
```

Comment:
I'm not sure I understand the reasoning with the precedences. Why are you using `ImplicitStyle` here?

Reply:
My reasoning here is that I need to make sure that RequestedTheme has precedence over the Inherited value - which ImplicitStyle does. The reason is that setting RequestedTheme somewhere in the tree "stops" the Foreground property inheritance

Reply:
Can you add a comment here explaining this?

I'm also curious how this would interact with issue of style vs inheritance precedence - https://github.com/unoplatform/uno/issues/3443

Reply:
Hmm, good question 🤔 . I am now realizing, I will probably also need a similar "trick" added to the `Control` foreground logic

Reply:
I was wrong, this is `TextBlock`-specific, it behaves a bit differently. I have updated the code to be less invasive about the implicit setting and all test I have done seem ok now. Let's see what UI test suite says. Regarding #3443, these changes, unfortunately, don't fix it, but it still behaves the same at least :-D . There is still the outstanding issue with visual states however (that theme is not properly applied when a button is pressed, etc.). If the tests pass, we could probably merge this and let that as a separate issue, as for existing scenarios there is no change.